PKCE configuration - enabled by default

rohan-naik07 opened the pull request:
Fixes [gh-16391](https://github.com/spring-projects/spring-security/issues/16391)

PKCE enabled by default for confidential as well as public clients.
Client Authentication method won't affect the PKCE customizer.
PKCE can be disabled using isRequireProofKey() client setting.

rohan-naik07 commented:
Ok, that makes sense. Thanks for clarifying my doubts.

jgrandja commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistration.java:
```diff
@@ -655,6 +655,10 @@ private ClientRegistration create() {
 			clientRegistration.clientName = StringUtils.hasText(this.clientName) ? this.clientName
 					: this.registrationId;
 			clientRegistration.clientSettings = this.clientSettings;
+			if (this.clientSettings.requireProofKey) {
```
Please remove this since `clientRegistration.clientSettings` is set in previous line with `this.clientSettings` and only the `clientSettings` should be used.

jgrandja commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistration.java:
```diff
@@ -756,9 +754,7 @@ public static final class ClientSettings implements Serializable {
-		private ClientSettings() {
-
-		}
+		private ClientSettings() {}
```
Please revert this formatting change as only required changes should be applied

jgrandja commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistration.java:
```diff
@@ -794,6 +790,7 @@ public static final class Builder {
 			private boolean requireProofKey;
 
 			private Builder() {
+				this.requireProofKey = true;
```
Please move this to variable initialization `private boolean requireProofKey = true;`

jgrandja commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolver.java:
```diff
@@ -184,8 +183,7 @@ private OAuth2AuthorizationRequest.Builder getBuilder(ClientRegistration clientR
 			}
-			if (ClientAuthenticationMethod.NONE.equals(clientRegistration.getClientAuthenticationMethod())
-					|| clientRegistration.getClientSettings().isRequireProofKey()) {
+			if (clientRegistration.getClientSettings().isRequireProofKey()) {
```
Public Clients (indicated with `ClientAuthenticationMethod.NONE`) always require PKCE but with this change it's possible to disable PKCE for Public Clients. Please revert this as the original logic is still correct.

jgrandja commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/DefaultServerOAuth2AuthorizationRequestResolver.java:
```diff
@@ -196,8 +195,7 @@ private OAuth2AuthorizationRequest.Builder getBuilder(ClientRegistration clientR
 			}
-			if (ClientAuthenticationMethod.NONE.equals(clientRegistration.getClientAuthenticationMethod())
-					|| clientRegistration.getClientSettings().isRequireProofKey()) {
+			if (clientRegistration.getClientSettings().isRequireProofKey()) {
```
Public Clients (indicated with `ClientAuthenticationMethod.NONE`) always require PKCE but with this change it's possible to disable PKCE for Public Clients. Please revert this as the original logic is still correct.

jgrandja commented on docs/modules/ROOT/pages/servlet/oauth2/client/client-authentication.adoc:
```diff
@@ -293,6 +293,6 @@ spring:
 ====
-Public Clients are supported using https://tools.ietf.org/html/rfc7636[Proof Key for Code Exchange] (PKCE).
-PKCE will automatically be used when `client-authentication-method` is set to "none" (`ClientAuthenticationMethod.NONE`).
+https://tools.ietf.org/html/rfc7636[Proof Key for Code Exchange] (PKCE) will be enabled by default for public as well as confidential clients. Refer https://docs.spring.io/spring-security/reference/servlet/oauth2/client/client-authentication.html#oauth2-client-authentication-public[this
```
Please revert this as the original content is still valid

jgrandja commented on docs/modules/ROOT/pages/servlet/oauth2/client/core.adoc:
```diff
@@ -69,7 +69,8 @@ This information is available only if the Spring Boot property `spring.security.
 The supported values are *header*, *form*, and *query*.
 <16> `userNameAttributeName`: The name of the attribute returned in the UserInfo Response that references the Name or Identifier of the end-user.
-<17> [[oauth2Client-client-registration-requireProofKey]]`requireProofKey`: If `true` or if `authorizationGrantType` is `none`, then PKCE will be enabled by default.
+<17> [[oauth2Client-client-registration-requireProofKey]]`requireProofKey`: This will by default be `true`, as PKCE will be enabled by default.
```
Please update the text to:

If `true` or if `clientAuthenticationMethod` is `none`, then PKCE will be enabled.

rohan-naik07 replied:
Aren't we stressing enough on the point that PKCE will be also enabled for confidential clients?... Or we don't want to for now

jgrandja commented on oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/registration/TestClientRegistrations.java:
```diff
@@ -64,6 +64,22 @@ public static ClientRegistration.Builder clientRegistration2() {
 		// @formatter:on
 	}
 
+	public static ClientRegistration.Builder publicClientRegistrationWithNoPkce() {
```
Please remove this as `TestClientRegistrations.clientRegistration()` can be used and the caller can simply change `builder.clientSettings(updatedClientSettings)`

jgrandja reviewed (changes requested):
Thanks for the PR @rohan-naik07. Please see review comments.

Also, please rebase off of `main` and ensure there is only 1 commit with the changes. Thanks.

rohan-naik07 commented:
Hi @jgrandja, i rebased my changes onto main branch hence some unwanted changes got pushed, sorry for inconvenience. I have removed all the changes and commented some doubts. Please look into it and update
